src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPathExtensions.cs

```diff
@@ -137,6 +137,22 @@ public static ODataPath TrimEndingTypeSegment(this ODataPath path)
             return handler.FirstPart;
         }
 
+        /// <summary>
+        /// Creates a <see cref="ODataPath"/> that is <paramref name="path"/> with the type segments and key segments removed from the end
+        /// </summary>
+        /// <param name="path">The <see cref="ODataPath"/> to trim the ending of</param>
+        /// <returns>A <see cref="ODataPath"/> without type-cast and key segments at the end</returns>
+        /// <exception cref="System.ArgumentNullException">Thrown if <paramref name="path"/> is <see langword="null"/></exception>
+        public static ODataPath TrimEndingTypeAndKeySegments(this ODataPath path)
+        {
+            if (path == null)

```

Comment:
shall we call the 'TrimEndingTypeSegment().TrimEndingKeySegment()' in this new method to avoid duplication?

Reply:
This is actually the bug that the PR is trying to fix. We don't want to trim the type segments and then trim the key segments. We want to trim any type and key segments. With the unit tests in the PR, you can try replacing this method with your comment and see how those tests fail.